src/net_processing.cpp

```diff
@@ -3497,45 +3494,41 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
 
         if (pfrom.fSuccessfullyConnected) {
-            // Disconnect peers that send a SENDTXRCNCL message after VERACK.
             LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "sendtxrcncl received after verack from peer=%d; disconnecting\n", pfrom.GetId());
             pfrom.fDisconnect = true;
             return;
         }
 
-        if (!peer->GetTxRelay()) {
-            // Disconnect peers that send a SENDTXRCNCL message even though we indicated we don't
-            // support transaction relay.
+        if (!peer->GetTxRelay() || m_ignore_incoming_txs) {
             LogPrintLevel(BCLog::NET, BCLog::Level::Debug, "sendtxrcncl received from peer=%d to which we indicated no tx relay; disconnecting\n", pfrom.GetId());
             pfrom.fDisconnect = true;
             return;
         }
```

Comment:
What about peers that send us relay=false in the VERSION message but we provide bloom filters to them. The code in this PR would not disconnect such a peer, but I think it should, right?

Reply:
These flags are confusing :(
Indeed, having `fRelay=false` I think is not currently compatible with reconciliations.

So, instead of `GetTxRelay()`, we better look at `pfrom.m_relays_txs`, right?

Reply:
> These flags are confusing :(

Yes! :frowning_face:

> Indeed, having `fRelay=false` I think is not currently compatible with reconciliations.

I guess you mean the relay flag in the either VERSION message - sent or received, right?

What about clarifying the BIP:

```diff
-sendtxrcncl ... Must not be sent if peer specified no support for transaction relay (fRelay=0) in "version". Otherwise, the sender should be disconnected.
+sendtxrcncl ... Must not be sent if either peer specified no support for transaction relay (fRelay=0) in "version". Otherwise, the sender should be disconnected.
```

> So, instead of `GetTxRelay()`, we better look at `pfrom.m_relays_txs`, right?

Yes, this is closer to checking whether the peer sent us `relay=false` in `VERSION`. However it is not identical - `fRelay` may be `true`, but `pfrom.m_relays_txs` be `false` if `pfrom.IsBlockOnlyConn()`. It is ok to accept this discrepancy because for block-only connections we sent `relay=false` in our `VERSION` message.

That said, maybe the condition should be:

```diff
-if (!peer->GetTxRelay() || m_ignore_incoming_txs) {
+if (!pfrom.m_relays_txs /* they sent relay=false */ ||
+    RejectIncomingTxs(pfrom) /* we sent relay=false */) {
```